### PR TITLE
User count metrics for site roles and special groups

### PR DESCRIPTION
**Pull request by labkey-jeckels**
#### Rationale
Knowing the numbers of users assigned to these site-wide roles and groups will help understand how they are used, and monitor for unusual assignments.

#### Changes
* New UsageMetricsProvider to capture user counts in site-wide roles and the special site admin and developer groups
* Minor code cleanup

### CODE REVIEW

File: api/src/org/labkey/api/security/SecurityManager.java [hunk 1]
```diff
@@ -182,6 +183,42 @@ public class SecurityManager
...
+                if (siteRoles.contains(role))
+                {
+                    int count = roleCounts.getOrDefault(role.getName(), 0);
+                    UserPrincipal principal = SecurityManager.getPrincipal(assignment.getUserId());
+                    if (principal != null && principal.isActive())
+                    {
+                        switch (principal.getPrincipalType())
+                        {
+                            case USER -> count++;
+                            case GROUP -> getAllGroupMembers((Group)principal, MemberType.ACTIVE_USERS);
```
Comment (labkey-adam) [inline api/src/org/labkey/api/security/SecurityManager.java]: Seems like a no-op. Don't you want to do something like `count += getAllGroupMembers().size()` here?

File: api/src/org/labkey/api/security/SecurityManager.java [hunk 2]
```diff
@@ -182,6 +183,42 @@ public class SecurityManager
         EmailTemplateService.get().registerTemplate(PasswordResetAdminEmailTemplate.class);
     }
 
+    public static UsageMetricsProvider getMetricsProvider()
+    {
+        return () -> {
+            Map<String, Object> result = new HashMap<>();
+
+            Map<String, Integer> roleCounts = new HashMap<>();
```
Comment (labkey-adam) [inline api/src/org/labkey/api/security/SecurityManager.java:191]: Optional: using a `HashBag` might be cleaner. It handles the initialization and accumulation, and provides `add(value, count)`.

**Review by labkey-adam — CHANGES_REQUESTED** [review]
Unless I'm misreading, the metrics provider group handling needs to be adjusted

File: api/src/org/labkey/api/security/SecurityManager.java (hunk 1 above)
Comment (labkey-jeckels) [reply api/src/org/labkey/api/security/SecurityManager.java]: Indeed

File: api/src/org/labkey/api/security/SecurityManager.java (hunk 2 above)
Comment (labkey-jeckels) [reply api/src/org/labkey/api/security/SecurityManager.java:191]: I didn't see an easy way to get the role name -> count map from it so I left as-is

**Review by labkey-adam — APPROVED** [review]

**Pull request merged**